### ISSUE TITLE
`infinite-loop`  `-> ! ` does not work if the loop is conditional

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16155]
Using the following flags

`--force-warn clippy::infinite-loop`

this code:

```rust
fn loopy(cond: bool) {
    let true = cond else { loop {} };
}

pub fn main() {}

```

caused the following diagnostics:

```
    Checking _a v0.1.0 (/tmp/icemaker_global_tempdir.TaFSR1bWYaDL/icemaker_clippyfix_tempdir.xGB7eUcT4zPY/_a)
warning: infinite loop detected
 --> src/main.rs:2:28
  |
2 |     let true = cond else { loop {} };
  |                            ^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#infinite_loop
  = note: requested on the command line with `--force-warn clippy::infinite-loop`
help: if this is intentional, consider specifying `!` as function return
  |
1 | fn loopy(cond: bool) -> ! {
  |                      ++++

warning: `_a` (bin "_a") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.17s

```

However after applying these diagnostics, the resulting code:

```rust
fn loopy(cond: bool) -> ! {
    let true = cond else { loop {} };
}

pub fn main() {}

```
no longer compiled:

```
    Checking _a v0.1.0 (/tmp/icemaker_global_tempdir.TaFSR1bWYaDL/icemaker_clippyfix_tempdir.xGB7eUcT4zPY/_a)
error[E0308]: mismatched types
 --> src/main.rs:1:25
  |
1 | fn loopy(cond: bool) -> ! {
  |    -----                ^ expected `!`, found `()`
  |    |
  |    implicitly returns `()` as its body has no tail or `return` expression
  |
  = note:   expected type `!`
          found unit type `()`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `_a` (bin "_a") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_a` (bin "_a" test) due to 1 previous error

```

Version:
```
rustc 1.93.0-nightly (3ff30e7ea 2025-11-29)
binary: rustc
commit-hash: 3ff30e7eafc1da7104c3960187d17939172428ed
commit-date: 2025-11-29
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5

```
